Add  read_and_decode API based iis3dwb sensor driver

avisconti opened the pull request:
The IIS3DWB is a ST 3-axis digital vibration sensor. This PR includes support using [read_and_decode APIs](https://docs.zephyrproject.org/latest/hardware/peripherals/sensor/read_and_decode.html) for both one shot and streaming capability. The current triggers supported are:

        - SENSOR_TRIG_FIFO_WATERMARK
        - SENSOR_TRIG_FIFO_FULL
        - SENSOR_TRIG_DATA_READY

This driver DOES NOT provide support for old [fetch_and_get APIs](https://docs.zephyrproject.org/latest/hardware/peripherals/sensor/fetch_and_get.html), therefore following ```sensor_driver_api``` callbacks (as well as trigger's threads support) are not provided:

```
   sensor_trigger_set_t trigger_set;
   sensor_sample_fetch_t sample_fetch;
   sensor_channel_get_t channel_get;
```

The driver has been tested on a nucleo_h503rb board with a spi-configured x_nucleo_iks01a3 shield. It should be in line with what discussed in #70651

EDIT:
rebased/updated/retested using the helper defined in #91775
Also added rtio_flush_completion_queue() new RTIO API which may be used to flush the completion queue.

avisconti committed:
drivers/sensor: iis3dwb: add streaming capabality

Add read_and_decode streaming APIs support.

Triggers supported:
    - SENSOR_TRIG_FIFO_WATERMARK
    - SENSOR_TRIG_FIFO_FULL
    - SENSOR_TRIG_DATA_READY

Signed-off-by: Armando Visconti <armando.visconti@st.com>

yperess commented on drivers/sensor/st/iis3dwb/iis3dwb.c:
```diff
@@ -0,0 +1,453 @@
+		odr = IIS3DWB_XL_ODR_OFF;
+		break;
+
+	default:
+		odr = IIS3DWB_XL_ODR_26k7Hz;
```
Other drivers only allow rounding up. We should keep that up, if the user asked for 25kHz then serving a higher value is OK, but if the user asks for 100kHz we should fail.

yperess commented on drivers/sensor/st/iis3dwb/iis3dwb.c:
```diff
@@ -0,0 +1,453 @@
+		}
+	}
+
+	if (edata->has_accel == 0) {
```
Won't this fail on SENSOR_CHAN_DIE_TEMP only reads? Do we have to read acceleration to get the die temperature?

yperess commented on drivers/sensor/st/iis3dwb/iis3dwb.c:
```diff
@@ -0,0 +1,453 @@
+
+static DEVICE_API(sensor, iis3dwb_driver_api) = {
+	.attr_set = iis3dwb_attr_set,
+#ifdef CONFIG_SENSOR_ASYNC_API
```
I think you should use `CONFIG_IIS3DWB_STREAM` here

yperess commented on drivers/sensor/st/iis3dwb/iis3dwb.c:
```diff
@@ -0,0 +1,453 @@
+			edata->has_accel = 1;
+
+			uint8_t outx_regs[]  = { IIS3DWB_OUTX_L_A, };
+			struct spi_buf buf_xl[] = { {(uint8_t *)edata->accel, 6} };
```
We shouldn't be using the memory from the driver here, the user should be providing the data with the SQE. If we have the MPU on, this would cause a fault.

yperess reviewed (dismissed):
Haven't finished a full review, but this should be a good start

avisconti commented on drivers/sensor/st/iis3dwb/iis3dwb.c:
```diff
@@ -0,0 +1,453 @@
+		read_reg = rtio_sqe_acquire(rtio);
+
+		if (write_addr == NULL || read_reg == NULL) {
+			return;
```
ack

avisconti commented on drivers/sensor/st/iis3dwb/iis3dwb.c:
```diff
@@ -0,0 +1,453 @@
+
+	complete_op = rtio_sqe_acquire(rtio);
+	if (complete_op == NULL) {
+		return;
```
ack

avisconti commented on drivers/sensor/st/iis3dwb/iis3dwb.c:
```diff
@@ -0,0 +1,453 @@
+	rc = rtio_sqe_rx_buf(iodev_sqe, min_buf_len, min_buf_len, &buf, &buf_len);
+	if (rc != 0) {
+		LOG_ERR("Failed to get a read buffer of size %u bytes", min_buf_len);
+		return;
```
ack

avisconti commented on drivers/sensor/st/iis3dwb/iis3dwb.c:
```diff
@@ -0,0 +1,453 @@
+	 *  Restore default configuration
+	 */
+	iis3dwb_reset_set(ctx, PROPERTY_ENABLE);
+	do {
+		iis3dwb_reset_get(ctx, &rst);
+	} while (rst);
```
ok, I'm looking at its usage

teburd commented:
> @yperess Thanks for your review. I'll take care of it as soon as possible. I think this work is important to consolidate and divulgate the read_and_decode approach. At this review stage I also want to ask you a comment about the iis3dwb_rtio_rd_transaction() routine. If you see among his argument there is a spi_buf structure pointer which basically it is used to pass a list of bufp and buf_len. It should be possible to prepare a RTIO buffer chunks reading, e.g. (using same concept in another driver I'm currently developing)
> 
> ```
> 		uint8_t outx_regs[]  = { LSM6DSVXXX_WHO_AM_I, LSM6DSVXXX_OUTX_L_A };
> 		struct spi_buf buf_xl[] = { {&edata->chip_id, 1} , {(uint8_t *)edata->accel, 6}, };
> 
> 		/*
> 		 * Prepare rtio enabled bus to read LSM6DSVXXX_OUTX_L_A register
> 		 * where accelerometer data is available.
> 		 * Then lsm6dsvxxx_one_shot_complete_cb callback will be invoked.
> 		 *
> 		 * STMEMSC API equivalent code:
> 		 *
> 		 *   uint8_t accel_raw[6];
> 		 *
> 		 *   lsm6dsvxxx_acceleration_raw_get(&dev_ctx, accel_raw);
> 		 */
> 		lsm6dsvxxx_rtio_rd_transaction(dev,
> 						outx_regs,
> 						ARRAY_SIZE(outx_regs),
> 						buf_xl,
> 						iodev_sqe,
> 						lsm6dsvxxx_one_shot_complete_cb);
> ```
> 
> Now I would like to know if
> 
>     1. Would such routine be useful for everybody and hence include it as a new API in rtio.h?
> 
>     2. Instead of using spi_buf (iis3dwb supports only SPI) we may use a different/custom styructure. We can define in rtio.h a new one.
> 
> 
> If this can be a good idea I may implement it in this PR at this stage. Let me know.
> 
> EDIT: Question open to everybody. Maybe @teburd may also have a comment



> @yperess Thanks for your review. I'll take care of it as soon as possible. I think this work is important to consolidate and divulgate the read_and_decode approach. At this review stage I also want to ask you a comment about the iis3dwb_rtio_rd_transaction() routine. If you see among his argument there is a spi_buf structure pointer which basically it is used to pass a list of bufp and buf_len. It should be possible to prepare a RTIO buffer chunks reading, e.g. (using same concept in another driver I'm currently developing)
> 
> ```
> 		uint8_t outx_regs[]  = { LSM6DSVXXX_WHO_AM_I, LSM6DSVXXX_OUTX_L_A };
> 		struct spi_buf buf_xl[] = { {&edata->chip_id, 1} , {(uint8_t *)edata->accel, 6}, };
> 
> 		/*
> 		 * Prepare rtio enabled bus to read LSM6DSVXXX_OUTX_L_A register
> 		 * where accelerometer data is available.
> 		 * Then lsm6dsvxxx_one_shot_complete_cb callback will be invoked.
> 		 *
> 		 * STMEMSC API equivalent code:
> 		 *
> 		 *   uint8_t accel_raw[6];
> 		 *
> 		 *   lsm6dsvxxx_acceleration_raw_get(&dev_ctx, accel_raw);
> 		 */
> 		lsm6dsvxxx_rtio_rd_transaction(dev,
> 						outx_regs,
> 						ARRAY_SIZE(outx_regs),
> 						buf_xl,
> 						iodev_sqe,
> 						lsm6dsvxxx_one_shot_complete_cb);
> ```
> 
> Now I would like to know if
> 
>     1. Would such routine be useful for everybody and hence include it as a new API in rtio.h?
> 
>     2. Instead of using spi_buf (iis3dwb supports only SPI) we may use a different/custom styructure. We can define in rtio.h a new one.
> 
> 
> If this can be a good idea I may implement it in this PR at this stage. Let me know.
> 
> EDIT: Question open to everybody. Maybe @teburd may also have a comment

I think if I'm reading this correctly this API is wanting to do a series of register reads, e.g. read reg X, Y, and Z into values X, Y, Z?

I mean this is completely reasonable, and actually I think matches some ideas @ubieda had around a regmap like API on top of RTIO.

I'd welcome rtio_regmap.h or some such that provides an API like this to read/write to a register map over a bus easily and with the ability to control blocking/non-blocking like all other RTIO stuff.

teburd reviewed (dismissed):
What a well written driver

Looking again the the bits here to read the interrupt status register and such though, I think this is very close to what @ubieda was thinking about. I'll try and suggest we discuss this in the next sensor wg if that makes sense to you @avisconti

avisconti commented:
This PR might be changed according to what decision will be taken on #91775,

avisconti committed:
drivers/sensor/st: add support to IIS3DWB accel

The IIS3DWB is a system-in-package featuring a 3-axis digital vibration
sensor with low noise over an ultrawide and flat frequency range.
The wide bandwidth, low noise, very stable and repeatable sensitivity,
together with the capability of operating over an extended temperature
range (up to +105 C), make the device particularly suitable for vibration
monitoring in industrial applications.

Datasheet: https://www.st.com/en/mems-and-sensors/iis3dwb.html

This driver is currently only supporting the polling-mode read_and_decode
APIs (both blocking and non-blocking).

This driver is based on stmemsc HAL i/f v2.9.1.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti commented:
Any feedback here guys?

avisconti commented on drivers/sensor/st/iis3dwb/iis3dwb.c:
```diff
@@ -0,0 +1,393 @@
+static void iis3dwb_one_shot_complete_cb(struct rtio *ctx, const struct rtio_sqe *sqe, void *arg)
+{
+	struct rtio_iodev_sqe *iodev_sqe = (struct rtio_iodev_sqe *)sqe->userdata;
+	int err = 0;
```
Sure, it is meant to get the return value of ```rtio_flush_completion_queue()``` below :(
Thanks for catching it!!

alexapostolu reviewed (commented):
This looks really good and it matches closely with the `lis2dux12` sensor driver too

avisconti commented on drivers/sensor/st/iis3dwb/iis3dwb.c:
```diff
@@ -0,0 +1,393 @@
+	}
+
+	if (iis3dwb_set_odr_raw(dev, odr)) {
+		LOG_DBG("failed to set sampling rate");
```
Yes sure, it makes sense. Ill change it, thx.

avisconti commented on drivers/sensor/st/iis3dwb/iis3dwb_rtio.h:
```diff
@@ -0,0 +1,19 @@
+
+#include <zephyr/device.h>
+#include <zephyr/rtio/rtio.h>
+#include <zephyr/rtio/work.h>
```
Usually I do not pay too much attention in avoiding including not necessary  .h files. I think it is a good idea removing it as it may lead to unnecessary compilations if this file change. And yes, the .h is not needed indeed!

I'm removing it.
Thanks

avisconti commented:
@yperess you requested changes which I already implemented some time ago. Do you happen to have some time to review my changes? Thanks
